language/diem-tools/prefetched-transaction-replay/src/lib.rs

```diff
@@ -0,0 +1,410 @@
+    #[structopt(name = "replay-transactions")]
+    ReplayTransactions { start: Version, limit: u64 },
+
+    /// Replay `account` transactions starting from version `start` to `start + limit`.
+    #[structopt(name = "replay-account-transactions")]
+    ReplayAccountTransactions {
+        account: AccountAddress,
+        start: Version,
+        limit: u64,
+    },
+
+    /// Execute serialized transactions from `txs_path`
+    #[structopt(name = "execute-serialized-transactions")]
+    ExecuteSerializedTransactions { txs_path: PathBuf },
+}
+
+pub struct PrefetchTransactionReplayer {
+    config: PrefetchTransactionReplayerConfig,
+    validator: Box<dyn DiemValidatorInterface>,
+    debugger: DiemDebugger,
+    rpc_client: BlockingClient,
+}
+
+impl StateView for OnDiskStateViewReader {
+    fn get(&self, ap: &AccessPath) -> Result<Option<Vec<u8>>> {
+        match ap.get_path() {
+            access_path::Path::Code(mid) => self.view.get_module_bytes(&mid),
+            access_path::Path::Resource(tag) => self.view.get_resource_bytes(ap.address, tag),
+        }
+    }
+
+    fn is_genesis(&self) -> bool {
+        false
+    }
+}
+
+fn ensure_dir_does_not_exist(dir: &Path) -> Result<()> {
+    if dir.exists() {
+        Err(format_err!("{:?} must be empty as it will be created", dir))
+    } else {
+        Ok(())
+    }
+}
+
+impl PrefetchTransactionReplayer {
+    pub fn from_config(config: PrefetchTransactionReplayerConfig) -> Result<Self> {
+        ensure_dir_does_not_exist(&config.build_dir)?;
+        ensure_dir_does_not_exist(&config.storage_dir)?;
+
+        let debugger = DiemDebugger::json_rpc_with_config(
+            &config.rpc_url,
+            config.build_dir.clone(),
+            config.storage_dir.clone(),
+        )?;
+
+        let rpc_client = BlockingClient::new(&config.rpc_url);
+
+        // NOTE: we could probably avoid creating a second validator
+        // by making the one in `DiemDebugger` accessible publicly
+        let validator = Box::new(JsonRpcDebuggerInterface::new(&config.rpc_url)?);
+        Ok(PrefetchTransactionReplayer {
+            config,
+            validator,
+            debugger,
+            rpc_client,
+        })
+    }
+
+    fn get_keys_read_for_tx(
+        &self,
+        live_state: &impl MoveResolver,
+        tx: &Transaction,
+        rw_analysis: &ReadWriteSetAnalysis,
+    ) -> Result<Vec<ResourceKey>> {
+        let signed_tx = match tx {
+            Transaction::UserTransaction(u) => u,
+            t => {
+                warn!("cannot process tx {:?}, ignoring", t);
+                return Ok(vec![]);
+            }
+        };
+
+        // NOTE: we might need to do something about the transactions we do
+        // not currently handle
+        match signed_tx.payload() {
+            TransactionPayload::ScriptFunction(_) => {
+                rw_analysis.get_keys_read(signed_tx, live_state)
+            }
```

Comment:
Are there any underlying issues why non-`ScriptFunction` payloads (transaction scripts in particular) can't be handled here?

Reply:
This is a limitation coming from rw_analysis I think. We didn't implement the read write set analysis for the old Scripts so the best thing we can do here is to invoke the shim layer to do the extra translation, but we won't be able to handle arbitrary scripts with the current analysis pass.